chapters/arrays.tex

```diff
@@ -1268,10 +1268,19 @@ \subsection{Element-wise Division}\label{array-element-wise-division}\label{elem
 
 \subsection{Element-wise Exponentiation}\label{element-wise-exponentiation}
 
-Exponentiation \lstinline!a ^ b! is defined as \lstinline[language=C]!pow(double a, double b)! in the ANSI~C library if both \lstinline!a! and \lstinline!b! are
-\lstinline!Real! scalars. A \lstinline!Real! scalar value is returned.  If \lstinline!a! or \lstinline!b! are \lstinline!Integer! scalars, they are
-automatically promoted to \lstinline!Real!.  Consequences of exceptional situations, such as ($\text{\lstinline!a!} = 0.0$ and $\text{\lstinline!b!} \leq 0.0$,
-$\text{\lstinline!a!} < 0$ and \lstinline!b! is not an integer) or overflow are undefined.
+Exponentiation \lstinline!a ^ b! always returns a \lstinline!Real! scalar value, and it is required that \lstinline!a! and \lstinline!b! are scalar \lstinline!Real! or \lstinline!Integer! values.
+The result should correspond to mathematical exponentiation (ideally the correctly rounded result based on the inputs) with the following special cases:
+\begin{itemize}
+\item If $\text{\lstinline!a!} = 0.0$ and $\text{\lstinline!b!} = 0$ for an \lstinline!Integer! expression \lstinline!b! the result is $1.0$.
+\item If $\text{\lstinline!a!} < 0$ and \lstinline!b! is an \lstinline!Integer! or a \lstinline!Real! having an integer value, the result is defined as $\pm |a|^b$, with sign depending on whether \lstinline!b! is even (positive) or odd (negative).
```

Comment:
Is it necessary to allow the `Real` numbers with zero fractional part here?  The special dependency on certain values of a `Real` expression seems messy to deal with, in particular if we think of the interactions with a unit system.  (This concern was also raised when I brought up this PR internally at Wolfram MathCore.)

The clean alternative would be:
```suggestion
\item If $\text{\lstinline!a!} < 0$ and \lstinline!b! is a non-zero \lstinline!Integer!, the result is defined as $\pm |a|^b$, with sign depending on whether \lstinline!b! is even (positive) or odd (negative).
\item When \lstinline!b! is \lstinline!Real! it is required that $\text{\lstinline!a!} > 0$.
```

Then I think one could make it even better by combining both rules for an `Integer` exponent into a single item:
> When \lstinline!b! is \lstinline!Integer! the result is defined by repeated multiplication of \lstinline!a!; for a non-negative \lstinline!b! it is equivalent to \lstinline!product(fill(a, b))!, for a non-positive \lstinline!b! it is equivalent to \lstinline!1 / product(fill(a, -b))!, and note that the two cases agree when \lstinline!b! is zero.

Reply:
> Is it necessary to allow the `Real` numbers with zero fractional part here? The special dependency on certain values of a `Real` expression seems messy to deal with, in particular if we think of the interactions with a unit system. (This concern was also raised when I brought up this PR internally at Wolfram MathCore.)

I would say that there is: backward compatibility, and compatibility with other languages (like C) I believe that Real numbers with zero fractional part makes sense.

However, I don't think we need to extend this special treatment to units.

To me the key insight is that the reason for making `0^0` undefined is that there is no unique limit when x and y go towards zero. However, if we know that the exponent is an integer the discussion about limit isn't relevant; the exponent is either zero or it isn't.

Reply:
And I don't think we should formally define `a^b` using repeated multiplication for some reasons:

- It is neither efficient (it is faster to compute `a^8` as `((a^2)^2)^2`) nor numerically optimal (round-off issues become a mess).
- If the base is also an integer we must first convert it to a real to avoid overflow
- It seems as is `a^4` and `a^4.0` are different - without any relation to each other. Thus, if we anyway need to explain that it is practically the same it seems better to define that it is the same - and then give a different non-normative explanation.

I agree that we can informally use it as non-normative text; and for symbolic manipulations it is useful.

Reply:
Yes, perhaps _corresponds to_ is better than _is defined by_?
> When \lstinline!b! is \lstinline!Integer! the result corresponds to repeated multiplication of \lstinline!a!, with a negative \lstinline!b! handled as \lstinline!1 / (a ^ (-b))!.
>\begin{nonnormative}
> For a non-negative \lstinline!b! this makes \lstinline!a ^ b! conceptually equivalent to \lstinline!product(fill(a, b))!, for a non-positive \lstinline!b! it is conceptually equivalent to \lstinline!1 / product(fill(a, -b))!, and note that the two cases agree when \lstinline!b! is zero.
Implementations should avoid actually performing repeated multiplication for reasons of numerics and performance.
> \end{nonnormative}

Reply:
> > Is it necessary to allow the `Real` numbers with zero fractional part here? The special dependency on certain values of a `Real` expression seems messy to deal with, in particular if we think of the interactions with a unit system. (This concern was also raised when I brought up this PR internally at Wolfram MathCore.)
> 
> I would say that there is: backward compatibility, and compatibility with other languages (like C) I believe that Real numbers with zero fractional part makes sense.

It is not that I have a problem with `Real` numbers with zero fractional part per se (especially not having worked on 32 bit platforms in the past), but that it is problematic that the semantics of the operation depends on whether the fractional part is zero or not.  Why don't we test this to see if we can find any library that performs exponentiation of a non-positive base and a `Real` exponent?

> However, I don't think we need to extend this special treatment to units.

For units, I believe one would need something even stronger: When `a` has non-empty unit in `a ^ b`, it is required that `b` is a constant `Integer` expression.

> To me the key insight is that the reason for making `0^0` undefined is that there is no unique limit when x and y go towards zero. However, if we know that the exponent is an integer the discussion about limit isn't relevant; the exponent is either zero or it isn't.

Ture, but I'm not sure it helps us resolve the more complicated questions.

Reply:
Specifically `(-a)^b` is `a^b*exp(i*pi*b)=a^b*(cos(b*pi)+i*sin(b*pi))`, for real `a>0` and `b`. And that is only real when `sin(b*pi)=0`, which happens when `b` has zero fractional part.

Reply:
> One could, of course, prohibit `(-a)^b` for Real `b` even when it has a well-defined Real result for zero fractional parts - but to me that seems like an unnecessary restriction.

Rather than an unnecessary restriction, I see it as a way of leveraging the type system to make errors detectable at translation time rather than at runtime.

Reply:
> > One could, of course, prohibit `(-a)^b` for Real `b` even when it has a well-defined Real result for zero fractional parts - but to me that seems like an unnecessary restriction.
> 
> Rather than an unnecessary restriction, I see it as a way of leveraging the type system to make errors detectable at translation time rather than at runtime.

I don't see how we leverage anything. Both checking whether a is negative or whether b has zero fractional parts is normally done at runtime, since we cannot always check it at translation time - and it is the combination that is problematic.

Introducing a hard error for this doesn't make sense; it just breaks compatibility for no obvious reason.

Reply:
We are leveraging the type system if we say it's safe to compute the power of any base as long as the exponent is an `Integer`.  Conversely, when the exponent is a `Real` we'd get a clean rule saying that the base must be positive (regardless the value of the exponent).

To me the two operations are different animals, and it is a perfect application of the type system to tell them apart.

Reply:
I just noticed that the suggestion contains "non-zero" as well. That (have been and) should only be applied for the _Real_ exponents with integer values since integer exponent should allow zero exponent, e.g., `(-2)^0` should be` |2|^0=1`. (There are many other ways of formulating this, e.g., the first item could generalize to say that integer exponent 0 gives `a^0=1.0`, regardless of the value for `a`.)